### PR TITLE
Use a single SSL context object

### PR DESCRIPTION
**Pull request by eakraly**
openssl allows multiple TLS version support through a single SSL_CTX object.

This PR replaces 4 per-version SSL_CTX objects with a single object (DTLS is not yet changed).
SSL context initialization code for openssl with modern API (>=1.1.0) uses  `TLS_server_method` and `SSL_CTX_set_min_proto_version` instead of enabling specific TLS version. Byproduct of this is TLSv1_3 support when used with openssl-1.1.1 and above

TLS 1.2 and TLS 1.3 cannot be disabled (as before)

Test plan:
- run_tests.sh script now runs turnserver with SSL certificate (which enables TLS support)
- run_tests.sh now has one more basic test that uses TLS protocol

### CODE REVIEW

File: examples/run_tests.sh [hunk 1]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 
 echo 'Running turnserver'
-../bin/turnserver --use-auth-secret  --static-auth-secret=secret --realm=north.gov --allow-loopback-peers --no-cli > /dev/null &
+../bin/turnserver --use-auth-secret  --static-auth-secret=secret --realm=north.gov --allow-loopback-peers --no-cli --cert ../examples/ca/turn_server_cert.pem --pkey ../examples/ca/turn_server_pkey.pem > /dev/null &
```
Comment (eakraly) [inline examples/run_tests.sh:4]: Run turnserver with SSL certificate to support TLS

File: examples/run_tests.sh [hunk 2]
```diff
@@ -16,6 +16,15 @@ else
 	exit $?
 fi
 
+echo 'Running turn client TLS'
```
Comment (eakraly) [inline examples/run_tests.sh:19]: Run uclient with TLS to test basic TLS support

File: src/apps/relay/mainrelay.c [hunk 3]
```diff
@@ -83,13 +83,7 @@ char HTTP_ALPN[128] = "http/1.1";
 #define DEFAULT_GENERAL_RELAY_SERVERS_NUMBER (1)
 
 turn_params_t turn_params = {
-NULL, NULL,
```
Comment (eakraly) [inline src/apps/relay/mainrelay.c:86]: Instead 4 SSL_CTX objects we now have 1

File: src/apps/relay/mainrelay.c [hunk 4]
```diff
@@ -3148,27 +3142,6 @@ static void set_ctx(SSL_CTX** out, const char *protocol, const SSL_METHOD* metho
 	{
 		int op = 0;
 
-#if defined(SSL_OP_NO_SSLv2)
```
Comment (eakraly) [inline src/apps/relay/mainrelay.c:3151]: Support for TLS versions here was duplicated
Only set in one place (so removing here)

**Review by ggarber — COMMENTED** [review]
Good job!

If understand it correctly this PR makes 2 backwards incompatible changes:

1/ SSL_OP_NO_SSLv2 is not disabled anymore.   It looks like in some old openssl versions it was supported and needed.

2/ Before it was possible to disable TLS 1.2 but keep 1.1 and 1.3 enabled and with this change that's not supported anymore.  I'm not saying that was useful, just trying to confirm if the behaviour changed and how the doc needs to be updated correspondingly if we really want to do that.

**Comment by eakraly** [comment]
Thank you!

1/ This is correct. Let me add this - I missed the point in older version of openssl.
2/ Correct. I think this is the right way to go (treating it as minimal TLS version and there is no way around it). +1 to adding documentation.

**Pull request merged**